src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -378,13 +410,8 @@ public Solution WithProjectOutputFilePath(ProjectId projectId, string? outputFil
         {
             CheckContainsProject(projectId);
 
-            var newState = _state.WithProjectOutputFilePath(projectId, outputFilePath);
-            if (newState == _state)
-            {
-                return this;
-            }
-
-            return new Solution(newState);
+            var newCompilationState = _compilationState.WithProjectOutputFilePath(projectId, outputFilePath);
```

Comment:
i started with that form, but i ended up really not liking it.  Ther was something extremely nice about everything being very closely consistent here.  It also helped avoid mistakes i found.  It could read slightly odd here.  LIke "why is the proejct output file path affecting the compilation state".  But it could also be read as: "the compilation state is allowed to know this is happening, and can decide if it needs to do anything here.

Given that the compilation-state *does* point at solution-state at the very least, it always at least makes some sense to have this pattern so you can get the new compilation state.

-

this also keeps the number of patterns low and the code simple and cosnsitent.